core/templates/dev/head/pages/exploration_player/AudioPreloaderService.js

```diff
@@ -20,117 +20,112 @@ oppia.factory('AudioPreloaderService', [
-      AudioTranslationManagerService, LanguageUtilService) {
-    // List of languages that have been preloaded in the exploration.
-    var _preloadedLanguageCodes = [];
+      AudioTranslationManagerService, LanguageUtilService,
+      ComputeGraphService) {
+    var MAX_NUM_AUDIO_FILES_TO_DOWNLOAD_SIMULTANEOUSLY = 3;
 
-    var _preloadAllAudioFiles = function(languageCode) {
-      var allAudioTranslations =
-        ExplorationPlayerStateService
-          .getExploration().getAllAudioTranslations(languageCode);
+    var _filesBeingDownloaded = [];
+    var _filesToBeDownloaded = [];
+    var _exploration = null;
+    var _isRunning = false;
 
-      allAudioTranslations.map(function(audioTranslation) {
-        AssetsBackendApiService.loadAudio(
-          ExplorationContextService.getExplorationId(),
-          audioTranslation.filename);
-      });
-
-      _preloadedLanguageCodes.push(languageCode);
+    var _init = function(exploration) {
+      _exploration = exploration;
     };
 
-    var _showBandwidthConfirmationModal = function(
-        audioTranslationsForContent, languageCode,
-        confirmationCallback) {
-      $modal.open({
-        templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
-          '/pages/exploration_player/' +
-          'audio_preload_bandwidth_confirmation_modal_directive.html'),
-        resolve: {},
-        backdrop: true,
-        controller: [
-          '$scope', '$modalInstance',
-          'ExplorationPlayerStateService', 'AudioPreloaderService',
-          'LanguageUtilService',
-          function(
-              $scope, $modalInstance,
-              ExplorationPlayerStateService, AudioPreloaderService,
-              LanguageUtilService) {
-            $scope.fileSizeOfCurrentAudioTranslationMB =
-              audioTranslationsForContent[languageCode]
-                .getFileSizeMB().toPrecision(3);
-            $scope.totalFileSizeOfAllAudioTranslationsMB =
-              ExplorationPlayerStateService.getExploration()
-                .getAllAudioTranslationsFileSizeMB(languageCode)
-                .toPrecision(3);
-            $scope.currentLanguageDescription =
-              LanguageUtilService.getAudioLanguageDescription(languageCode);
-            $scope.shouldDownloadAllAudioInExploration = false;
+    var _bfsForAudioFiles = function(sourceStateName) {
+      var languageCode = AudioTranslationManagerService
+        .getCurrentAudioLanguageCode();
+      var explorationGraph = ComputeGraphService.compute(
+        _exploration.getInitialState().name, _exploration.getStates());
+      var queue = [];
+      var seen = {};
+      var audioFiles = [];
+      allAudioTranslations = _exploration.getAllAudioTranslations(
+        languageCode);
+      queue.push(sourceStateName);
+      while (queue.length > 0) {
+        var currStateName = queue.shift();
+        var audioFile = allAudioTranslations[currStateName];
+        if (audioFile != null) {
```

Comment:
Perhaps `if (!allAudioTranslations.hasOwnProperty(currStateName))`? (and drop the previous line)

Reply:
Eh I prefer it the way it is (it gets used on the next line too)

Reply:
Ah wait, hang on, I think I'm confused. allAudioTranslations is specified by the exploration, right? If so, why would this be null? Wouldn't the state name always appear as a key?

(Also we should use == and not ===.)

Reply:
Oh, I was thinking that if there's a state with no audio translation, then that could result in a null value, right?

And sorry why should we use == instead of ===?

Reply:
Oh sorry I misspoke completely! I meant, we should use === instead of == (and !== instead of !=). The triple one checks types as well.

Shouldn't that case result in an empty list/dict?

Reply:
Ohh, I see.

No I think it should be null, the individual values within allAudioTranslation are AudioTranslation domain objects. Just realized I should change the variable name from audioFile to audioTranslation.

Reply:
Oh, I think I get it now. But in that case I would still suggest using hasOwnProperty rather than trying to find a key in the dict that doesn't exist (and rely on the default value being null/undefined). As in, if the point is that currStateName may not exist in the allAudioTranslations dict, we should check for that non-existence rather than try to resolve it to a value (unless we always expect the key to exist and are setting null explicitly somewhere).

Reply:
P.S. The latter approach is fine too, it just has to be one or the other. I also suggest documenting getAllAudioTranslations() to be clear about what exactly is being returned.